src/Hamiltonians/geometry.jl

```diff
@@ -37,12 +41,104 @@ julia> geo[(3,3)]
 julia> geo[(3,4)] # returns 0 if out of bounds
 0
 ```
+"""
+abstract type Geometry{D} end
+
+Base.size(g::Geometry, i) = size(g)[i]
+Base.length(g::Geometry) = prod(size(g))
+
+"""
+    num_neighbors(::Geometry)
+
+Return the number of neighbors each site has in the geometry.
+"""
+num_neighbors
+
+"""
+    neighbor_site(::Geometry, site, i)
+
+Find the `i`-th neighbor of `site` in the geometry. If the move is illegal, return 0.
+"""
+neighbor_site
+
+"""
+    periodic_dimensions(::Geometry{D})
+
+Return a `D`-tuple of `Bool`s signaling which dimensions of the [`Geometry`](@ref) are
+periodic.
+"""
+periodic_dimensions
+
+"""
+    num_dimensions(geom::Geometry)
+
+Return the number of dimensions of the lattice in this geometry.
+"""
+num_dimensions(::Geometry{D}) where {D} = D
+
+function Base.getindex(g::Geometry{D}, vec::Union{NTuple{D,Int},SVector{D,Int}}) where {D}
+    return get(LinearIndices(size(g)), fold_vec(g, SVector(vec)), 0)
+end
+Base.getindex(g::Geometry, i::Int) = SVector(Tuple(CartesianIndices(size(g))[i]))
+
+"""
+    fold_vec(g::Geometry{D}, vec::SVector{D,Int}) -> SVector{D,Int}
+
+Use the [`Geometry`](@ref) to fold the `vec` in each dimension. If folding is disabled in a
+dimension, the vector is allowed to go out of bounds.
```

Comment:
Confuses me a bit. What does it mean to fold?

Is the following interpretation accurate?
```suggestion
Use the [`Geometry`](@ref) to fold the `vec` in each dimension back into the bounded domain. 
If folding is disabled in a dimension, the vector is unchanged and allowed to go out of bounds.
```